### ISSUE TITLE
Suggestion: Passing Args to Callback in Flux.train!, and return args from Loss()

### ISSUE THREAD

**Issue by cems2** [FluxML/Flux.jl#999]
**Problem:**  Flux.train allows for a callback but without any args.  This creates unweildy interfaces requiring global variables to work around.  A different but, as it turns out, related problem occurs with the loss functions return args.  In some problems the loss function computes expensive sub calculations one would like to access without redoing-- again demanding global variables.  

**Rationale:**
It turns out that solving these in a combined way is better than ad hoc independent solutions to both.  Rather than some rigid definition of what values get passed to the call-back we will use a free byproduct of the user-defined loss function call to create a user-defined set of args for the user-defined callback.  

**Suggested resolution:**
Allow the loss function to return a tuple with two args.  The first arg is the standard scalar loss required by Flux to optimize.  The second arg is any user-defined payload-- for example, a tuple of intermediate calculations in the loss functions such as the last output of the network and separate loss terms like regularizers.  

This second arg is to be ignored by Flux.train! itself, but will be passed as the arg to the designated callback function.

Example:
```
function loss(x)   **# this is a typical looking loss function, nothing special at all**
   pred = prediction(x.value)
   loss1  = abs2.( pred - x.labels)
   r = regularizers(x.network)
   totalLoss = sum( loss1) + r
   payload = ( loss1, pred, r)
  return totalLoss, payload
end

function callback(totalLoss, payload)  # the callback takes an arg
   println ("totalLoss", totalLoss ) 
   println ("Current Regulaizer value ",payload[3]
   println ("Current total Loss ", payload[1] )
   Plot( payload[2] )  **# plot the details of the prediction without recomputing it!** 
   # maybe some decision criteria to save a copy of the best loss or stopping criteria
end

# the loss train! function retains the same interface.
Flux.train!( loss, parameters, dataset, opt; cb = callback )

```

**maintaining backward compatibility:**
All the new train! function has to do is detect if loss returned a single scalar or a tuple.  If a tuple it extracts the loss for it's own use and then passes the tuple  to callback. if a scaler it calls the callback with no args.

**Comment by CarloLucibello** [comment]
This proposal seem an improvement to me, but honestly, I would discourage everywhere (flux's docs and model-zoo) the whole `train!` approach in favor of plain for loops  

```julia
for d in data
    gs = gradient(ps) do
      training_loss = loss(d...)
      # Insert what ever code you want here that needs Training loss, e.g. logging
      return training_loss
    end
    # insert what ever code you want here that needs gradient
    # E.g. logging with TensorBoardLogger.jl as histogram so you can see if it is becoming huge
    update!(opt, ps, gs)
    # Here you might like to check validation set accuracy, and break out to do early stopping
end
```

**Comment by cems2** [comment]
Hey thanks for that code.  The reason most people use train I think is that unless you know the internals the process pitfalls of writing your own are obscure for new folks.

The suggestion was recently adopted by diffeqflux and obviates using train!

Thus having the same command in flux too seems logical .

In diffeqflux the new train loop is
Diffeqflux.sciml_train()

And the proposed change allows the use of BFGS in the  train loop.  Huge speed increase on polishing results!!!

**Comment by CarloLucibello** [comment]
Closing as addressed.

The callback half is moot: `Flux.train!` no longer accepts `cb` at all — it errors and points to a manual `gradient`/`update!` loop.

The more interesting half of this proposal — letting the loss return a payload alongside the scalar — *is* implemented. The loss may return `(loss, aux...)`, and `Flux.trainstep!` returns that value, so intermediate computations from the loss are available without globals or recomputation.
